### PR TITLE
Put bucket encryption using kms key alias from app-data.yml

### PR DESCRIPTION
**Pull request by andrewpearce-digital**
Does not create a KMS key.
Only puts encryption when default_encryption is present in app-data.yml.

### CODE REVIEW

**Comment by andrewpearce-digital** [comment]
At this stage, a KMS key is being manually created. If this works, the next step is to also create a KMS key for s3 for each stack

File: roles/ec2-s3-bucket/tasks/sub-create.yml [hunk 1]
```diff
@@ -44,3 +44,25 @@
             --bucket {{ bucket_name }}  \
             --lifecycle-configuration file://lifecycle.json "
   when: "'lifecycle' in bucket_data and 'non_current_version_expiration' in bucket_data.lifecycle"
+
+- name: Get KMS Key metadata
+  command: "aws kms describe-key --key-id alias/{{ default_encryption_kms_keyname }}"
+  register: kms_metadata
+  check_mode: no
+  when: "'default_encryption_kms_keyname' in bucket_data"
```
Comment (carinadigital) [inline roles/ec2-s3-bucket/tasks/sub-create.yml]: Check this condition carefully.
Do you need a default value? What happens if YAML Key default_encryption_kms_keyname is empty?

**Review by carinadigital — COMMENTED** [review]
You will need to check this on other stacks too before merging.
Pity this isn't part of work to move off ubuntu 14.04 to be done in other tools.

File: roles/ec2-s3-bucket/tasks/sub-create.yml (hunk 1 above)
Comment (andrewpearce-digital) [reply roles/ec2-s3-bucket/tasks/sub-create.yml]: The intention is that if default_encryption_kms_keyname is empty, this should be skipped, no encryption applied to the bucket.
There should be no default value for the YAML key

**Comment by andrewpearce-digital** [comment]
> You will need to check this on other stacks too before merging.
> Pity this isn't part of work to move off ubuntu 14.04 to be done in other tools.

It's work towards that, it gets the app changed at the cost of investing in older tooling.

**Comment by andrewpearce-digital** [comment]
Now responds to this in `app-data.yml`
```
s3_buckets:
  - name: lpa-pdf-cache
    default_encryption_kms: false # << valid options are true, false, or "remove"
```
See https://opgtransform.atlassian.net/browse/LPA-3078

File: roles/ec2-s3-bucket/tasks/default-encryption-kms.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+- name: Get KMS Key metadata
+  command: "aws kms describe-key --key-id alias/{{ bucket_name }}"
+  register: kms_metadata
+  check_mode: no
+
+- name: Get KMS Key ID
+  set_fact:
+    kms_key_id: "{{ (kms_metadata.stdout | from_json | json_query('KeyMetadata.KeyId')) }}"
+  check_mode: no
```
Comment (carinadigital) [inline roles/ec2-s3-bucket/tasks/default-encryption-kms.yml:56]: should depend on above succeeding

File: roles/ec2-s3-bucket/tasks/default-encryption-kms.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+- name: Check for existing KMS key
+  command: "aws kms describe-key --key-id alias/{{ bucket_name }}"
+  register: kms_metadata
+  check_mode: no
+
+- name: Generate new KMS Key Policy
+  template:
+    src: kms-key-policy.json.j2
+    dest: "{{ playbook_dir }}/kms-key-policy.json"
+  when: kms_metadata.stderr != ""
```
Comment (carinadigital) [inline roles/ec2-s3-bucket/tasks/default-encryption-kms.yml]: comment please

Comment (carinadigital) [reply roles/ec2-s3-bucket/tasks/default-encryption-kms.yml]: Set fact for no key exists

**Review by carinadigital — CHANGES_REQUESTED** [review]

**Review by carinadigital — APPROVED** [review]

File: roles/ec2-s3-bucket/tasks/default-encryption-kms.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
+---
+
+- name: Check for existing KMS key
+  command: "aws kms describe-key --key-id alias/{{ bucket_name }}"
+  register: kms_metadata
+  check_mode: no
+
+# When no KMS key exists, stderr returns an empty string
```
Comment (carinadigital) [inline roles/ec2-s3-bucket/tasks/default-encryption-kms.yml]:
```suggestion
# When KMS key exists, stderr returns an empty string
```

**Review by carinadigital — CHANGES_REQUESTED** [review]

**Review by carinadigital — APPROVED** [review]

**Pull request merged**